clang/lib/CIR/Dialect/Transforms/TargetLowering/CIRContext.h

```diff
@@ -0,0 +1,74 @@
+//===- CIRContext.h - Context to lower CIR ----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Partially mimics AST/ASTContext.h. The main difference is that this is
+// adapted to operate on the CIR dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_LIB_CIR_DIALECT_TRANSFORMS_TARGETLOWERING_CIRCONTEXT_H
+#define LLVM_CLANG_LIB_CIR_DIALECT_TRANSFORMS_TARGETLOWERING_CIRCONTEXT_H
+
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Types.h"
+#include "clang/AST/Type.h"
+#include "clang/Basic/TargetInfo.h"
+#include "llvm/ADT/IntrusiveRefCntPtr.h"
+
+namespace mlir {
+namespace cir {
+
+// FIXME(cir): Most of this is type-related information that should already be
+// embedded into CIR. Maybe we can move this to an MLIR interface.
+class CIRContext : public llvm::RefCountedBase<CIRContext> {
+
+private:
+  mutable SmallVector<Type, 0> Types;
+
+  const clang::TargetInfo *Target = nullptr;
+  const clang::TargetInfo *AuxTarget = nullptr;
+
+  /// MLIR context to be used when creating types.
+  MLIRContext *MLIRCtx;
+
+  /// The language options used to create the AST associated with
+  /// this ASTContext object.
+  clang::LangOptions &LangOpts;
+
+  //===--------------------------------------------------------------------===//
+  //                         Built-in Types
+  //===--------------------------------------------------------------------===//
+
+  Type CharTy;
+
+public:
+  CIRContext(MLIRContext *MLIRCtx, clang::LangOptions &LOpts);
+  CIRContext(const CIRContext &) = delete;
+  CIRContext &operator=(const CIRContext &) = delete;
+  ~CIRContext();
+
+  /// Initialize built-in types.
+  ///
+  /// This routine may only be invoked once for a given ASTContext object.
+  /// It is normally invoked after ASTContext construction.
+  ///
+  /// \param Target The target
+  void initBuiltinTypes(const clang::TargetInfo &Target,
+                        const clang::TargetInfo *AuxTarget = nullptr);
+
+private:
+  void initBuiltinType(Type &R, clang::BuiltinType::Kind K);
```

Comment:
Assuming this is `mlir::Type`, does this need to be `&R` instead of `R`? Or are you expecting that to "return" the desired type? If so, should it properly return the type instead? Or is this related to being similar to OG codegen?

Reply:
It's intentional: a null type `R` is passed and initialized within the function. This is due to codegen parity, but it is a minor detail and returning the value sounds better. I'll update it.